### PR TITLE
make DimHints compatible with Dims

### PR DESCRIPTION
**Pull request by pianpwk**
Previously we'd been raising UserErrors when `Dim()` and DimHints (`Dim.AUTO/Dim.DYNAMIC`) were both specified in `dynamic_shapes`, this PR stops that, and uses `Dim()` objects to guide DimHints. 

The key to this was making the `EqualityConstraint` class happy when it checks that inferred equivalence relations were specified in the original `dynamic_shapes` spec, and this introduces a `RelaxedConstraint` object to mark the hinted dimensions, so equality checks between `RelaxedConstraints` and other constraints are treated as valid.

Current behavior is that:
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        return x - y

inputs = (torch.randn(4, 4), torch.randn(4, 4))
shapes = {
    "x": (Dim.AUTO, Dim("d1", min=3)),
    "y": (Dim("d0", max=8), Dim.DYNAMIC),
}
ep = export(Foo(), inputs, dynamic_shapes=shapes)
```

The dimensions marked `AUTO` and `DYNAMIC` will have max & min ranges of 8 & 3 respectively. Note that inferred equality between `Dim()` objects & `Dim.STATIC` will still raise errors - `Dim()` suggests not specializing to a constant. 

Differential Revision: D64636101

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Review by avikchaudhuri — APPROVED** [review]
Can you confirm the following?
(1) inferred equalities are still recorded / asserted as such in the graph
(2) AUTO and STATIC / constant dims are compatible

**Comment by pianpwk** [comment]
> Can you confirm the following? (1) inferred equalities are still recorded / asserted as such in the graph (2) AUTO and STATIC / constant dims are compatible

Yep! Verified locally, and I think there's a handful of test cases that check this in some form

**Comment by wdvr** [comment]
@pytorchbot merge -i

merging since it landed internally

**Comment by wdvr** [comment]
@pytorchbot merge -f

merging since it landed internally
